improve Blosc compression operator

psychocoderHPC opened the pull request:
fix: #2018

The current implementation of the Blosc ADIOS2 operator is due to limitations of Blosc not supporting adios variables larger than 2GiB.
ADIOS1 was originally using a chunked format to work around this limitation.

- Add option `threshold` to set a threshold under which data will only be copied instead of compressed (equivalent to Blosc in ADIOS1).
- Use chunk format to avoid 2GiB limit for variables (datasets).
- Add support to decompress Blosc data written with ADIOS before this PR.
- - update blosc tests

The header written to decompressed the date is different from the ADIOS1 implementation. 
Compared to ADIOS1 this PR implements compression and decompression much cleaner.

@maintainer Where do I should document the new option `threshold` for the Blosc operator. I have not found anything in the documentation.

Thanks to @franzpoeschel for providing the test code [testBlosc.txt](https://github.com/ornladios/ADIOS2/files/5842414/testBlosc.txt)
.
CC-ing: @ax3l

pnorbert commented:
Thank you for fixing this compressor. Unfortunately, you caught us with the docs, there is no section on operators/compressors. 
@chuckatkins Can you please add redirect this to the patch release?

williamfgc commented:
@psychocoderHPC thanks, that was long overdue, only bzip2 was doing that and the hope was to migrate to blosc2 eventually as in [this issue](https://github.com/Blosc/c-blosc/issues/67#issuecomment-263234555). @pnorbert there is almost nothing in the [docs](https://adios2.readthedocs.io/en/latest/components/components.html#operator) on compression operations since each engine supports its own set.

psychocoderHPC commented:
> @psychocoderHPC Can you please extend the tests in `testing/adios2/engine/bp/operations/TestBPWriteReadBlosc.cpp` so that the threshold settings are tested? You need to look at this test anyway as the memory sanitizer fails on it...

I updated the blosc tests and added test combinations for `shuffling` and `threshold`

psychocoderHPC commented on source/adios2/operator/compress/CompressBlosc.h:
```diff
@@ -60,6 +64,50 @@ class CompressBlosc : public Operator
+                               void *dataOut, const size_t sizeOut,
+                               Params &info) const;
+
+    class __attribute__((packed)) DataHeader
```
@pnorbert I fixed hopefully the `ubsan` issue in the CI by disallowing the compiler to add padding. 
The disadvantage is that there is no standard attribute in C++ :-(
Is this way ok? If not, do you have a suggestion on how I can fix this issue instead?

williamfgc replied:
if it helps, there is an [internal helper function](https://github.com/ornladios/ADIOS2/blob/master/source/adios2/helper/adiosMemory.h#L244) to add padding in a portable way.

psychocoderHPC replied:
I am not sure if this helps. The problem is that maybe the ptr for the output is not aligned during the compression. If I use the helper and shift my data few bytes and later during decompress I get the data within a array with a different base address. If so it would be not possible to find the data because the required padding is different.

If I understand the ubsan error for the old code correctly and what I found in the net the packed attribute is required to guarantee that the struct has the same layout even if you compile blosc on a different system. This attributes forbits the compiler to transform the struct which is IMO correct.

I can also remove the header class und work with two pure integers but I found it better readable to use the header class because it allowes in the code to shift with `sizeOf(Header)` instead using a magic number 8byte.

I am not sure which solution guarantee portability.

psychocoderHPC replied:
Maybe it is enough if I align the struct to 4 byte.
Is ADIOS guaranteeing that my pointers given to compress operators are 4 byte aligned?

williamfgc replied:
> The problem is that maybe the ptr for the output is not aligned during the compression. 

The function I shared can verify that returning the required padding in bytes (if 0 no padding is required).

> Is ADIOS guaranteeing that my pointers given to compress operators are 4 byte aligned?

That's a good question. It depends on the serialization buffer consuming this interface. The BP buffer is `char` so it shouldn't be a problem. I'd check with @pnorbert for other existing strategies.

> I am not sure which solution guarantees portability.

Yup, that's the challenge.

ax3l replied:
It looks like Blosc is not covered by CI and thus breaking on MSVC: #2745

psychocoderHPC commented:
>  that and the hope was to migrate to blosc2 eventually as in [this issue](https://github.com/Blosc/c-blosc/issues/67#issuecomment-263234555).

I also hoped to switch to blosc2 and it is solving the 2GiB issue but unluckily as I know they now support larger size-types but you need to explicitly pack the data with the framing API and there is no simple function available which is doing it transparently.
Therefore I decided to stay with the well-tested approach from ADIOS1 with blosc1.

chuckatkins commented:
I've rebased the topic on release_27 as a candidate for the 2.7.1 patch release.  Please update your local branch with the remote.
